### DISCUSSION TITLE
Questions about configs in launch_java.sh

### DISCUSSION THREAD

**Discussion by LinzeSu** [strimzi/strimzi-kafka-operator#8553] [Q&A] [answered]
Hi,
I noticed that in file launch_java.sh, there're configs:
https://github.com/strimzi/strimzi-kafka-operator/blob/b0a575f87ddc9dd0dde573c2d5f55840a69ed019/docker-images/operator/scripts/launch_java.sh#L15
and:
https://github.com/strimzi/strimzi-kafka-operator/blob/b0a575f87ddc9dd0dde573c2d5f55840a69ed019/docker-images/operator/scripts/launch_java.sh#L26-L29
I assume the goal of these two configs is that the app does not encounter the problem of oom cause it's running in container. My questions are:
1. Does those two settings assure that no OOM problem will occur?
2. The setting -XX:MaxRAMPercentage=20 seems very conservative. Is there any practical guidance of this setting?

The questions might not be related to the actual use of the strimzi or kafka. Any insights are appreciated.

**Comment by scholzj** [comment, marked as answer]
> Does those two settings assure that no OOM problem will occur?

The aim is to minimize OoM occurrences. But I don't think you can really be 100% sure. There are also various use cases - some users have one Kafka cluster, while some have tens or hundreds of different resources. So it is hard to find one setting to fit them all.

> The setting -XX:MaxRAMPercentage=20 seems very conservative. Is there any practical guidance of this setting?

IIRC this controls the amount of RAM which will go to the Java heap. That is only a small chunk of the memory used by a Java application. So it looks like very little, but the other parts of memory take a fairly big chunk of it when you give the process only something like 384Mi by default. The script allows users to tune these numbers when needed -> so for example if you decide to give the operator 1Gi of memory because you run many clusters, it might be possible to tune this number as well.

**Reply by LinzeSu** [reply]
Hi scholzj,

Thank you for the response. You said:
> So it looks like very little, but the other parts of memory take a fairly big chunk of it when you give the process only something like 384Mi by default.

To my knowlegde there could be off-heap memory(GC, jvm itself, metadata and etc.) and other process that also use memory. I built a streaming app with pure java which reads from and writes to kafka, and it keeps exceeding the pod memory limit, slowly but definitely. The `jcmd app_pid VM.native_memory summary` shows that the used memory(sum of committed heap and off-heap) does not change much after a couple of hours(which should indicate that there's no memory leak), but the `top` command shows the memory usage is slowly increasing. I set the heap size to use part of the total memory like in the script did but it still failed. Now I'm testing the app using MALLOC_ARENA_MAX=2  according to post described [here](https://siddhesh.in/posts/malloc-per-thread-arenas-in-glibc.html) and just found out that this script also did what I did.

I noticed that [someone](https://stackoverflow.com/a/73832632/19476830) also encountered this problem. Just wonder if you know why the RSS memory usage showed in top is higher than the total committed memory showed using jcmd command. Again, might not be the best place to ask but I'm kind of lost in looking for solutions.

**Reply by scholzj** [reply]
I'm afraid I'm not that much of an expert in this area - and they way I got to the 20%. was mostly experimenting and seeing what fails or does not fail :-/.

**Reply by LinzeSu** [reply]
Got it. Still, thanks for answering.
